File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/MaxSubArraySum.java

import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;

class MaxSubArraySum {
    
    /*@
    public normal_behavior
      ensures \result == ((x % 4294967296L + 6442450944L) % 4294967296L) - 2147483648L;
    model public static pure \bigint fb_wrap(\bigint x);

        public normal_behavior
          requires a != null && 0 <= k <= a.length;
          ensures \result == (k == 0 ? 0 : fb_wrap(fb_tail(a, k - 1) + a[k - 1]) < 0 ? 0 : fb_wrap(fb_tail(a, k - 1) + a[k - 1]));
          measured_by k;
        model public static pure \bigint fb_tail(int[] a, int k);
        public normal_behavior
          requires a != null && 0 <= k <= a.length;
          ensures \result == (k == 0 ? 0 : fb_best(a, k - 1) >= fb_wrap(fb_tail(a, k - 1) + a[k - 1]) ? fb_best(a, k - 1) : fb_wrap(fb_tail(a, k - 1) + a[k - 1]));
          measured_by k;
        model public static pure \bigint fb_best(int[] a, int k);
        public normal_behavior
          requires a != null && 0 <= k <= a.length;
          ensures \result == (k == 0 ? 0 : fb_wrap(fb_tail(a, k - 1) + a[k - 1]) < 0 ? k : fb_start(a, k - 1));
          measured_by k;
        model public static pure int fb_start(int[] a, int k);
        public normal_behavior
          requires a != null && 0 <= k <= a.length;
          ensures \result == (k == 0 ? 1 : fb_best(a, k - 1) < fb_wrap(fb_tail(a, k - 1) + a[k - 1]) ? k - fb_start(a, k - 1) : fb_length(a, k - 1));
          measured_by k;
        model public static pure int fb_length(int[] a, int k);

        public normal_behavior
          requires size <= 0 || (a != null && size <= a.length);
          assignable \nothing;
          ensures \result == (size <= 0 ? 1 : fb_length(a, size));
          also
        public exceptional_behavior
          requires size > 0 && a == null;
          assignable \nothing;
          signals_only NullPointerException;
          also
        public exceptional_behavior
          requires a != null && size > a.length;
          assignable \nothing;
          signals_only ArrayIndexOutOfBoundsException;
    @*/
    //@ code_java_math
    public static int maxSubArraySum(int[] a, int size) {
        int max_so_far = 0;
        int max_ending_here = 0;
        int start = 0;
        int end = 0;
        int s = 0;

        /*@
        loop_invariant 0 <= i <= (size < 0 ? 0 : size);
        loop_invariant 0 <= s <= i;
        loop_invariant 0 <= start <= end && end <= (i == 0 ? 0 : i - 1);
        loop_invariant a != null && i <= a.length ==> max_ending_here == fb_tail(a, i) && max_so_far == fb_best(a, i) && s == fb_start(a, i) && end - start + 1 == fb_length(a, i);
        loop_writes i, max_so_far, max_ending_here, start, end, s;
        decreases size <= 0 ? 0 : (\bigint)size - i;
        @*/
        for (int i = 0; i < size; i++) {
            max_ending_here += a[i];
            if (max_so_far < max_ending_here) {
                max_so_far = max_ending_here;
                start = s;
                end = i;
            }
            if (max_ending_here < 0) {
                max_ending_here = 0;
                s = i + 1;
            }
        }
        return end - start + 1; // Returns the length of the maximum sum subarray
    }
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/MaxSubArraySum.c
#include "java_arrays.h"
#include <stdint.h>

/*@
assigns \nothing;
ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
assigns \nothing;
ensures \result == (int32_t)((integer)left - right);
*/
static int32_t java_sub(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left - (uint32_t)right);
}

/*@
logic integer fb_tail{L}(JIntArray a, integer k) = k <= 0 ? 0 : \max(0, (int32_t)(fb_tail(a, k - 1) + a->data[k - 1]));
logic integer fb_best{L}(JIntArray a, integer k) = k <= 0 ? 0 : \max(fb_best(a, k - 1), (int32_t)(fb_tail(a, k - 1) + a->data[k - 1]));
logic integer fb_start{L}(JIntArray a, integer k) = k <= 0 ? 0 : (int32_t)(fb_tail(a, k - 1) + a->data[k - 1]) < 0 ? k : fb_start(a, k - 1);
logic integer fb_length{L}(JIntArray a, integer k) = k <= 0 ? 1 : fb_best(a, k - 1) < (int32_t)(fb_tail(a, k - 1) + a->data[k - 1]) ? k - fb_start(a, k - 1) : fb_length(a, k - 1);
*/
/*@
requires size <= 0 || (jintarray_valid_nonnull(a) && size <= a->length);
assigns \nothing;
ensures \result == (size <= 0 ? 1 : fb_length(a, size));
*/
int32_t maxSubArraySum(JIntArray a, int32_t size) {
    int32_t max_so_far = 0;
    int32_t max_ending_here = 0;
    int32_t start = 0;
    int32_t end = 0;
    int32_t s = 0;

    /*@
    loop invariant 0 <= i <= \max(0, size);
    loop invariant 0 <= s <= i;
    loop invariant 0 <= start <= end <= \max(0, i - 1);
    loop invariant size > 0 ==> max_ending_here == fb_tail(a, i) && max_so_far == fb_best(a, i) && s == fb_start(a, i) && end - start + 1 == fb_length(a, i);
    loop assigns i, max_so_far, max_ending_here, start, end, s;
    loop variant \max(0, size) - i;
    */
    for (int32_t i = 0; i < size; i = java_add(i, 1)) {
        max_ending_here = java_add(max_ending_here, jarray_get(a, i));
        if (max_so_far < max_ending_here) {
            max_so_far = max_ending_here;
            start = s;
            end = i;
        }
        if (max_ending_here < 0) {
            max_ending_here = 0;
            s = java_add(i, 1);
        }
    }
    return java_add(java_sub(end, start), 1);
}
